✨ Added methods to intercept calls to a fake k8s client

ludydoo opened the pull request:
This PR adds some methods to allow consumers of the fake client to intercept kubernetes api calls to inject custom logic, which is helpful while testing e.g. expected behavior upon certain types of failures.

alvaroaleman commented:
/ok-to-test

ludydoo commented:
Hi @alvaroaleman , thanks a lot for your quick review! What's your opinion about: would it make more sense to put this into the `client` package? Personally I don't see much use-case for a non-testing scenario, so the `fake` package makes sensse, but in terms of usage, perhaps it's clearer if consumers will use some `client.Interceptor` rather than some `fake.Interceptor` ?

ludydoo commented on pkg/client/fake/intercept.go:
```diff
@@ -0,0 +1,176 @@
+	return c.client.DeleteAllOf(ctx, obj, opts...)
+}
+
+func (c interceptor) Status() client.SubResourceWriter {
```
@inteon good point, no benefit at all. I'll remove that.

inteon replied:
we then probably want to call `c.SubResource("status")` here instead of `c.client.Status()`?

ludydoo replied:
@inteon fixed and added tests for that.

ludydoo commented:
@alvaroaleman @inteon please let me know if there is anything I should tweak or modify

alvaroaleman commented:
@ludydoo we had a lot of discussion about fakeclient changes like this, which is why I can't just merge without some of the other maintainers approving and two of them are on PTO unfortunately. So this will take a bit of time, but there is nothing expected from you at this point. Sorry about that.

vincepri reviewed (commented):
This looks like a good way to allow some mocking to happen in the fake client, I'm personally +1, but would like to hear @sbueringer's opinion as well.

sbueringer commented:
Just back from PTO, will take a look ASAP, not sure if it will be before KubeCon

sbueringer commented:
Looks good to me as well. Looks like it needs a rebase to pick up the change to the golangci-lint action.

One thought (similar to: https://github.com/kubernetes-sigs/controller-runtime/pull/2248#issuecomment-1486245876): as this functionality is probably often used together with fake client, but it can also be used with a regular client, should we move it out of the fake package?

I was thinking about maybe a new package like pkg/client/intercepter. Has the benefit to keep it clearly separate from the fake client and if we rename the func we could make it e.g. interceptor.NewClient instead of fake.Interceptor.

alvaroaleman commented:
If we move this into a distinct package, can we make this discoverable by integrating with the fake client builder by adding a With method for each of the interceptor methods that sets them?

ludydoo commented:
> If we move this into a distinct package, can we make this discoverable by integrating with the fake client builder by adding a With method for each of the interceptor methods that sets them?

I'm not sure how to resolve the circular dependency in this case, since the `interceptor_test` import the `fake` package.

ludydoo commented:
/retest

sbueringer commented on pkg/client/interceptor/intercept.go:
```diff
@@ -0,0 +1,197 @@
+
+type (
+	// GetFn is a function that can be used to intercept GET calls.
+	GetFn func(ctx context.Context, client client.WithWatch, key client.ObjectKey, obj client.Object, opts ...client.GetOption) error
```
Was it necessary to have the funcs as separate types? (vs just inline in the struct as before)

ludydoo replied:
No, I thought it was cleaner this way, but I removed it as it is not necessary

sbueringer commented on pkg/client/interceptor/intercept.go:
```diff
@@ -0,0 +1,197 @@
+	SubResourcePatchFn func(ctx context.Context, client client.SubResourceClient, obj client.Object, patch client.Patch, opts ...client.SubResourcePatchOption) error
+
+	// Fns contains functions that are called instead of the underlying client's methods.
+	Fns struct {
```
Sorry missed this before. Can we maybe replace all Fn with Func and Fns with Funcs? (in all files)

Just reads a lot better in my opinion

ludydoo replied:
Done

sbueringer commented:
@ludydoo Can you please drop the merge commits & rebase to main?

(I know you probably kept them for easier reviews, but the delta of this PR would be easier to see without the merge commits)

ludydoo commented:
> @ludydoo Can you please drop the merge commits, squash and rebase to main?
> 
> (I assume you probably kept them for easier reviews, but the delta of this PR would be easier to see without the merge commits)

@sbueringer just addressed your comments, and squashed & rebased to main. Please let me know if you think there should be some more changes.

sbueringer commented:
@ludydoo Thank you very much!!

/lgtm

/assign @vincepri @alvaroaleman

alvaroaleman reviewed (approved):
Thanks!

alvaroaleman commented:
Thanks a lot for this @ludydoo ! This makes injecting custom beheavior into the fakeclient for tests much easier :)

ludydoo commented:
Thank you for your reviews and advices @alvaroaleman @sbueringer @inteon @vincepri !
